rfcs/20240531-coo-sparse-encoding/README.md

```diff
@@ -0,0 +1,173 @@
+# Introducing Co-ordinate Sparse Encodings (COO)
+
+## Objective
+The memory API in oneDNN is currently capable of handling sparse data operations but its sparsity support is limited to compressed sparse row (CSR) and packed sparse encodings only. The objective of this proposal is to extend the API to also include support for co-ordinate sparse encodings (COO).
+
+The extension is introduced in two steps:
+- defining updates to the memory descriptor API for COO encodings.
+- introducing a reference matmul implementation that processes COO-encoded sparse tensors.
+
+## Motivation
+Many deep learning (DL) workloads now employ sparse tensors for their operations due to the performance boost achieved from lower memory usage and reduced computational load. 
+But the storage schemas using which the data is sparsified can differ from application to application. 
+Mainstream DL frameworks incorporate sparsity support by selecting from various encoding formats such as CSR (Compressed Sparse Row), COO (Co-ordinate Sparse Encoding), CSF (Compressed Sparse Fibers) and  BSC (Blocked Compressed Sparse Column). 
+The memory API in oneDNN is designed to be scalable to support all the needed encodings but the currently supported ones are CSR and Packed Sparse Encodings. 
+The aim is to also extend support to additional sparse data formats beginning with the COO format which is commonly used in mainstream frameworks.
+
+## COO and CSR Encodings
+Both COO and CSR are hardware-agnostic sparse encodings meaning that the storage schema can be implemented irrespective of the hardware on which it is run. Implementation-wise, the COO format is a simpler, more generalized version of CSR.
+The CSR format sparsifies data using three buffers:
+
+|   | CSR Buffers    | Contents |
+|:--|:---------------|:---------|
+|0  | `data`	     | Non-zero matrix values sorted in row-major order |
+|1  | `indices`      | Column indices corresponding to each non-zero value | 
+|2  | `pointers`     | Compressed row indices - each $n^{th}$ value denotes the number of non-zero elements between $n^{th}$ and $(n+1)^{th}$ row |
+|||
+
+Because of the compressed row, this format tends to be more efficient for large sparse matrices.
+COO, on the other hand, is simpler in implementation in that the encoding comprises of a list of thruples `(values, row_index, column_index)` corresponding to the non-zero values.
+This makes the COO less efficient than CSR but it has the advantage of a reduced conversion overhead and better interpretability. 
+For practical cases, a sorted variant of COO is used wherein the data is encoded as a set of sorted arrays containing the values, row indices and column indices respectively:
```

Comment:
Do we explicitly require users to pass us the memory in the canonical COO format where the entries are sorted by row, then column?

Also, do we allow the values to be zero?

Reply:
Ideally, the generic COO format does not require the user to explicitly pass the memory in a row-major order. But this will be the case for sorted COO as it determines how the buffers are defined for the data indices. 
With the given declaration,  the indices for dimension $n$ will be stored in buffer index $[n+1]$, that is, buffer[1] will hold indices for dimension 0, buffer[2] will hold indices for dimension 1 and so on. To do this, a pre-defined order for the sorted entries will be necessary.
The value buffers are expected to only hold non-zero elements.